src/qga.rs

```diff
@@ -45,118 +30,71 @@ impl QgaWrapper {
+                Ok(s) => s,
+                Err(e) => {
+                    error!("Failed to connect QGA, retrying: {}", e);
+                    thread::sleep(Duration::from_secs(1));
+                    continue;
                 }
-                QgaWrapperCommand::GuestExec(args) => match qga.execute(&args) {
-                    Ok(r) => Ok(QgaWrapperCommandResp::GuestExec(r)),
-                    Err(e) => Err(anyhow!("Failed to guest_exec: {}", e)),
-                },
-                QgaWrapperCommand::GuestExecStatus(args) => match qga.execute(&args) {
-                    Ok(r) => Ok(QgaWrapperCommandResp::GuestExecStatus(r)),
-                    Err(e) => Err(anyhow!("Failed to guest_exec_status: {}", e)),
-                },
             };
-
-            // Note we do not care if this succeeds or not.
-            // It is OK if receiver has gone away (eg we got timed out).
-            let _ = send_resp.send(resp);
-        }
-    }
-
-    /// Ask the worker thread to complete a request
-    fn execute(&self, cmd: QgaWrapperCommand, timeout: Duration) -> Result<QgaWrapperCommandResp> {
-        match self.send_req.send(cmd) {
-            Ok(_) => (),
-            Err(e) => {
-                debug!("Failed to send QGA command: {}", e);
-                bail!("Failed to send QGA command: worker thread is dead")
-            }
-        };
-
-        match self.recv_resp.recv_timeout(timeout) {
-            Ok(r) => r,
-            Err(RecvTimeoutError::Timeout) => bail!("Timed out waiting for QGA"),
-            Err(RecvTimeoutError::Disconnected) => {
-                bail!("Failed to recv QGA command: worker thread is dead")
+            qga_stream.set_read_timeout(Some(Duration::from_secs(5)))?;
+            let mut qga = Qga::from_stream(&qga_stream);
+            let sync_value = rand::thread_rng().gen_range(1..10_000);
+            match qga.guest_sync(sync_value) {
+                Ok(_) => {
+                    return Ok(Self { stream: qga_stream });
+                }
+                Err(e) => {
+                    warn!("QGA sync failed, retrying: {e}");
+                    thread::sleep(Duration::from_secs(1));
+                }
             }
         }
-    }
 
-    /// Sync with qemu-ga inside guest
-    pub fn guest_sync(&self, timeout: Duration) -> Result<()> {
-        self.execute(QgaWrapperCommand::GuestSync, timeout)
-            .map(|_| ())
+        Err(anyhow!("Timed out waiting for QGA connection"))
     }
 
     /// Run a command inside the guest
     pub fn guest_exec(
         &self,
         args: qga::guest_exec,
     ) -> Result<<qga::guest_exec as QapiCommand>::Ok> {
-        match self.execute(QgaWrapperCommand::GuestExec(args), Duration::MAX)? {
-            QgaWrapperCommandResp::GuestExec(resp) => Ok(resp),
-            _ => panic!("Impossible return"),
+        if let Err(e) = self.stream.set_read_timeout(None) {
+            warn!("Error setting socket timeout: {e}");
         }
```

Comment:
Maybe we should set a 5s timeout for all qga calls? At least for the APIs we use, they're all async. I'd be surprised if any took seconds. Unless you've seen otherwise.

That would keep the timeout setting centralized. And remove the save/restore stuff below

Reply:
done